### PR TITLE
feat: store vnode bitmap in local state store

### PR DESCRIPTION
**Pull request by hzxa21**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR lets local state store aware of the vnodes it is responsible for by storing vnode bitmap in read version. The vnode bitmap will be initialized during state store init and updated on state table `update_vnode_bitmap`. This PR also adds an extra assertion on read version to ensure there is no staging uncommitted data on `update_vnode_bitmap`.

#15099 

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/src/mem_table.rs [hunk 1]
```diff
@@ -434,6 +436,7 @@ pub struct MemtableLocalStateStore<S: StateStoreWrite + StateStoreRead> {
     table_id: TableId,
     op_consistency_level: OpConsistencyLevel,
     table_option: TableOption,
+    vnodes: Option<Arc<Bitmap>>,
```
Comment (wenym1) [inline src/storage/src/mem_table.rs:439]: For simplicity I think it's unnecessary to store the vnode in this trivial local state store.

File: src/storage/src/hummock/store/local_hummock_storage.rs [hunk 2]
```diff
@@ -427,6 +436,19 @@ impl LocalStateStore for LocalHummockStorage {
...
             .expect("should be able to send")
     }
+
+    fn update_vnode_bitmap(&mut self, vnodes: Arc<Bitmap>) -> Arc<Bitmap> {
+        let mut read_version = self.read_version.write();
+        assert!(read_version.staging().is_empty(), "There is uncommitted staging data in read version table_id {:?} instance_id {:?} on vnode bitmap update",
+            self.table_id(), self.instance_id()
+        );
+        let prev_vnodes = read_version.update_vnode_bitmap(vnodes);
+        assert!(
```
Comment (wenym1) [inline src/storage/src/hummock/store/local_hummock_storage.rs:446]: Can be merge into `prev_vnodes.expect("Previous vnode bitmap should not be none")`

**Review by wenym1 — APPROVED** [review]
Rest LGTM!

File: src/storage/src/mem_table.rs (hunk 1 above)
Comment (hzxa21) [reply src/storage/src/mem_table.rs:439]: If we don't store vnodes in this local state store, I think the assertion in state table to check whether the state table and state store vnode bitmap matches will fail.

**Pull request merged**